Throwing an exception when the afterSave aborts the transaction

lorenzo opened the pull request:
Fixes #9079

lorenzo committed:
Throwing an exception when the afterSave aborts the transaction

Fixes #9079

ADmad committed:
Fix CS errors.

lorenzo committed:
Fixing tests and CS errors

markstory commented:
Seems reasonable to me. Will this new code result in an exception being raised if an afterSave() listener commits the parent transaction as well?

lorenzo commented:
@markstory Yes, but that is more difficult to do by accident, since you will need to commit as many times as `begin` was called... that cannot just happen by accident

markstory commented:
@lorenzo Fair point.

Xymanek commented:
Suggestion: add a parent database exception which controller or anything else that calls the save() can try/catch and display a nice "Database error" message

P.s. On mobile now, will provide an example if I'm not clear when I get home

markstory commented:
@lorenzo Fair point.

lorenzo commented:
@Xymanek already done that way :)

ionas commented on src/ORM/Exception/RolledbackTransactionException.php:
```diff
@@ -0,0 +1,25 @@
+class RolledbackTransactionException extends Exception
+{
+
+    protected $_messageTemplate = 'The afterSave event in "%s" is aborting the transaction before the save process is done.';
```
Is this a developer/internal error or something end users should see? If latter: How does it get translated? Can we simply add `__d('cake',` via constructor ..?

markstory replied:
If your end users see this, you goofed up imo.

ADmad commented on src/ORM/Exception/RolledbackTransactionException.php:
```diff
@@ -0,0 +1,25 @@
+use Cake\Core\Exception\Exception;
+
+/**
+ * Used when a behavior cannot be found.
```
This description needs to be updated :)

Xymanek commented:
@lorenzo I guess I wasn't clear then, my bad. What I'm looking for is the following:

```php
try {
	$table->save();
} catch (DatabaseException $e) {
	$this->Flash->error('Database error');
} finally {
	$this->Flash->success('Something has been saved!');
}
```

So far I've checked the `save()` and it throws a couple of RuntimeExceptions and InvalidArgumentException. The exception that you are adding right now extends base Cake's exception. They don't share a parent so I can't `catch() {}` them all in one statement.

P.S. There is also PDOException that needs to be caught separately, but still...
P.P.S. I just realized that this is outside of the scope of current PR, but I said it here because here is a new exception is added

lorenzo commented:
@Xymanek This is a developer error, though, not a database error.

Xymanek commented:
By DatabaseException I mean any exceptions raised in the model layer of the application. If you think that the name is misleading, how about ModelException or something?

markstory commented on src/ORM/Exception/RolledbackTransactionException.php:
```diff
@@ -0,0 +1,25 @@
+ * Used when a behavior cannot be found.
+ *
+ */
+class RolledbackTransactionException extends Exception
```
To @Xymanek's point we could extend `Cake\Database\Exception` instead?

thinkingmedia commented:
@Xymanek I agree.

Rename the exception to `TransactionException` so that it can be used again elsewhere, and extend `Cake\Database\Exception`.

lorenzo committed:
Rebasing and fixing docblock

lorenzo commented:
I rebased and fixed the docblock. On the topic of renaming the exception or making it extend another class: This is exception meant for development mode, geared towards preventing a developer error cause because of a limitation in the ORM. It is not an "exceptional" case caused by a database corrupt state (such as a duplicate key  exception).

The intention is that this exception does not get caught by accident and confused with any other possible database error. It need to be solved before the code goes to production.

ADmad commented:
LGTM

markstory commented:
Nice work @lorenzo
